include/pybind11/detail/smart_holder_type_casters.h

```diff
@@ -464,6 +464,24 @@ struct shared_ptr_trampoline_self_life_support {
     }
 };
 
+template <typename T,
+          typename D,
+          typename std::enable_if<std::is_default_constructible<D>::value, int>::type = 0>
+inline std::unique_ptr<T, D> unique_with_deleter(T *raw_ptr, std::unique_ptr<D> &&deleter) {
+    if (deleter != nullptr) {
+        return std::unique_ptr<T, D>(raw_ptr, std::move(*deleter));
```

Comment:
I could add a unit test with a custom deleter that is default constructible to ensure a unit test exists through this code path. Currently that's missing, I can do that tomorrow evening, but first want to know if we want the extra complexity introduced by this addition.

Reply:
I think the extra complexity is not a significant concern. Long-term, and when used at scale, completeness is more valuable.

Minor and optional: How about making this more similar to the other specialization?

```
if (deleter == nullptr) {
    return std::unique_ptr<T, D>(raw_ptr);
}
return std::unique_ptr<T, D>(raw_ptr, std::move(*deleter));
```

Yes to the extra unit test :-)

Reply:
I added the extra unit test to the latest push, the line was a bit long on the python side so that became a bit ugly to be both readable and under the line length.

If ci fails (likely) I'll fix that in the evening, off to work for now, the last commit before today ran clean on GHA, if you want to do an internal test run as well.